src/client/conn.rs

```diff
@@ -0,0 +1,456 @@
+                        Err(_) => panic!("dispatch dropped without returning error"),
+                    }
+                }))
+            },
+            Err(req) => {
+                debug!("connection was not ready");
+                let err = ::Error::new_canceled(None);
+                Either::B(future::err((err, Some(req))))
+            }
+        };
+        Box::new(inner)
+    }
+}
+
+/*
+impl<T, B> Service for SendRequest<T, B> {
+    type Request = Request<B>;
+    type Response = Response;
+    type Error = ::Error;
+    type Future = ResponseFuture;
+
+    fn call(&self, req: Self::Request) -> Self::Future {
+
+    }
+}
+*/
+
+impl<B> fmt::Debug for SendRequest<B> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("SendRequest")
+            .finish()
+    }
+}
+
+// ===== impl Connection
+
+impl<T, B> Connection<T, B>
+where
+    T: AsyncRead + AsyncWrite,
+    B: Stream<Error=::Error> + 'static,
+    B::Item: AsRef<[u8]>,
+{
+    /*
+    pub fn into_inner(self) -> T {
+
+    }
+    */
+
+    /// Return the inner IO object, and additional information.
+    pub fn into_parts(self) -> Parts<T> {
+        let (io, read_buf) = self.inner.into_inner();
+        Parts {
+            io: io,
+            read_buf: read_buf,
+            _inner: (),
+        }
+    }
+
+    /// Poll the connection for completion, but without calling `shutdown`
+    /// on the underlying IO.
+    ///
+    /// This is useful to allow running a connection while doing an HTTP
+    /// upgrade. Once the upgrade is completed, the connection would be "done",
+    /// but it is not desired to actally shutdown the IO object. Instead you
+    /// would take it back using `into_parts`.
+    pub fn poll_without_shutdown(&mut self) -> Poll<(), ::Error> {
+        self.inner.poll_without_shutdown()
+    }
+}
+
+impl<T, B> Future for Connection<T, B>
+where
+    T: AsyncRead + AsyncWrite,
+    B: Stream<Error=::Error> + 'static,
+    B::Item: AsRef<[u8]>,
+{
+    type Item = ();
+    type Error = ::Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        self.inner.poll()
+    }
+}
+
+impl<T, B> fmt::Debug for Connection<T, B>
+where
+    T: AsyncRead + AsyncWrite + fmt::Debug,
+    B: Stream<Error=::Error> + 'static,
+    B::Item: AsRef<[u8]>,
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("Connection")
+            .finish()
+    }
+}
+
+// ===== impl Builder
+
+impl Builder {
+    /// Creates a new connection builder.
+    #[inline]
+    pub fn new() -> Builder {
+        Builder {
+            h1_writev: true,
+            keep_alive: true,
+        }
+    }
+
+    pub(super) fn h1_writev(&mut self, enabled: bool) -> &mut Builder {
+        self.h1_writev = enabled;
+        self
+    }
+
+    /// Set whether to use keep-alive on this connection.
+    ///
+    /// Default is true.
+    #[inline]
+    pub fn keep_alive(&mut self, enabled: bool) -> &mut Builder {
```

Comment:
Does this inject a `Connection: close` or just cause poll to return ready after the first response comes in?

Reply:
Currently it tells the internal `h1::Conn` that after the first response, it cannot be used for a new request, and sets its state as `Closed`. This is the same as disabling `keep_alive` on the `Client` as well.

I've sometimes thought that it should probably set `Connection: close`, to be nice, but then, even if the header isn't set, once a response has been read, it's up to the client to use it again, or close the connection.

Interestingly, I think this option would do nothing if the connection were HTTP2. Or maybe it could send a `GOAWAY` after the request...

Reply:
Given that it wouldn't do anything for h2, would it make more sense to not have this method at all and have users set Connection: close to manage keep alive? Hyper does pay attention if that header is set on outbound requests I think, right?